src/main/java/io/supertokens/storage/mysql/Start.java

```diff
@@ -898,4 +902,305 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
+            throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            PasswordlessQueries.deleteDevicesByEmail_Transaction(this, sqlCon, email);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void createCode(PasswordlessCode code) throws StorageQueryException, UnknownDeviceIdHash,
+            DuplicateCodeIdException, DuplicateLinkCodeHashException {
+
+        try {
+            PasswordlessQueries.createCode(this, code);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            if (e.actualException instanceof UnknownDeviceIdHash) {
+                throw (UnknownDeviceIdHash) e.actualException;
+            }
+            String message = e.actualException.getMessage();
+            if (message.matches(
+                    ".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessCodesTable() + ".*PRIMARY\\'")) {
+                throw new DuplicateCodeIdException();
+            }
+            if (message.matches(".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessCodesTable()
+                    + ".*link\\_code\\_hash\\'"))
+                if (message.contains("link_code_hash")) {
+                    throw new DuplicateLinkCodeHashException();
+                }
+            throw new StorageQueryException(e.actualException);
+        }
+    }
+
+    @Override
+    public PasswordlessCode getCodeByLinkCodeHash_Transaction(TransactionConnection con, String linkCodeHash)
+            throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            return PasswordlessQueries.getCodeByLinkCodeHash_Transaction(this, sqlCon, linkCodeHash);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void deleteCode_Transaction(TransactionConnection con, String deviceIdHash) throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            PasswordlessQueries.deleteCode_Transaction(this, sqlCon, deviceIdHash);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void createUser(io.supertokens.pluginInterface.passwordless.UserInfo user) throws StorageQueryException,
+            DuplicateEmailException, DuplicatePhoneNumberException, DuplicateUserIdException {
+        try {
+            PasswordlessQueries.createUser(this, user);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            String message = e.actualException.getMessage();
+            if (message.matches(
+                    ".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessUsersTable() + ".*PRIMARY\\'") ||
```

Comment:
Not sure why are are using "*". Are you sure it won't give false positives? If yes, we should consider making other checks with "*" as well..?

Reply:
I am happy to discuss more about regular expressions, I really enjoy building them! It shouldn't give false positives, since it needs to contain "Duplicate entry passwordless_users PRIMARY" keywords in these sequence. But again, we can discuss and improve it.

Reply:
The sequence is guaranteed, but it still matches strings that the exception detection part of other methods won't. So while I think doing this in a single regex is good, we should keep the pattern the same.

Reply:
so from what I understand, we are not making any changes in this section right ?

Reply:
No, I meant we need to change it to match what we have originally or copy the old way of exception detection. Either one is fine with me; copying is better for consistency.

Reply:
understood, thanks!